src/ast/parse.rs

```diff
@@ -46,20 +46,54 @@ pub fn parse_expr(
 
 /// A [`Span`] contains the file name and a pair of offsets representing the start and the end.
 #[derive(Clone, PartialEq, Eq, Hash)]
-pub struct Span(Arc<SrcFile>, usize, usize);
+pub enum Span {
+    /// Panics if a span is needed. Prefer `Span::Rust` (see `span!`)
+    /// unless this behaviour is explicitly desired.
+    Panic,
+    /// A span from a `.egg` file.
+    /// Constructed by `parse_program` and `parse_expr`.
+    Egglog(Arc<EgglogSpan>),
+    /// A span from a `.rs` file. Constructed by the `span!` macro.
+    Rust(Arc<RustSpan>),
+}
+
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub struct EgglogSpan {
+    file: Arc<SrcFile>,
+    i: usize,
+    j: usize,
+}
+
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub struct RustSpan {
+    pub file: &'static str,
+    pub line: u32,
+    pub column: u32,
+}
 
-lazy_static::lazy_static! {
-    pub static ref DUMMY_SPAN: Span = Span(Arc::new(SrcFile {name: None, contents: String::new()}), 0, 0);
+#[macro_export]
+macro_rules! span {
+    () => {
+        $crate::ast::Span::Rust(std::sync::Arc::new($crate::ast::RustSpan {
+            file: file!(),
+            line: line!(),
+            column: column!(),
+        }))
+    };
 }
 
 impl Span {
     pub fn string(&self) -> &str {
-        &self.0.contents[self.1..self.2]
+        match self {
+            Span::Panic => panic!("Span::Panic in Span::String"),
+            Span::Rust(_) => todo!(),
```

Comment:
This would just load the file from disk and load that line number? Is the question about the column?

FWIW I know Python just went through a refactor process so that they now store starting line, staring column, as well as ending line and ending column for bytecdoe ops and tracebacks: https://peps.python.org/pep-0657/

Do you think that might make sense in this case or is there a reason that just starting line and column is more appropriate?

Reply:
Yes, the problem is that Rust only exposes `file!`, `line!`, and `column!`.

Docs: https://doc.rust-lang.org/std/macro.file.html